### PR TITLE
Implement AesCng and TripleDESCng

### PR DESCRIPTION
**Pull request by AtsushiKan**
Implement AesCng and TripleDESCng.

This code is derived from the internal CNG-based Aes implementation in SSC.Algorithms
but the factoring has been redone so that the padding logic, cipher algorithm and the
underlying provider are now properly separated.

UniversalCryptoEncryptor and UniversalCryptoDecryptor implement the public-facing
ICryptoTransform interface and do the padding/depadding.

They are parameterized by a BasicSymmetricCipher which implements the unpadded transform
and encapsulates the cipher algorithm and underlying crypto technology (CNG or OpenSSL or
whatever.)

BasicSymmetricCipherCng is the CNG-based implementation of that, and
is parameterized by the algorithm (AES or 3DES.)

(The existing Aes implementations in Algorithms could be rewritten in terms of these
but I'm not doing that now. We don't even know that we'll keep those inside
SSC.Algorithms, in which case, redoing them would be throwaway work. I've put the agnostic
code in the Common folder just in case.)

Beefed up the shared Aes tests with a few more code-coverage tests, but what's there
seems pretty comprehensive already.

TripleDES tests exercise the basic cipher. No need to do more code-coverage with these
as AesCng and TripleDES are the virtually the same class on the managed side.

cc @bartonjs

### CODE REVIEW

File: src/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            return decryptedBytes;
+        }
+
+        protected sealed override byte[] UncheckedTransformFinalBlock(byte[] inputBuffer, int inputOffset, int inputCount)
+        {
+            // We can't complete decryption on a partial block
+            if (inputCount % InputBlockSize != 0)
+                throw new CryptographicException(SR.Cryptography_PartialBlock);
+
+            byte[] outputData;
```
Comment (stephentoub) [inline src/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs]: Nit: this should be declared down closer to where it's actually set.

File: src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            outer.BaseKeySize = keySize;
+            _keyName = null; // Setting _keyName to null signifies that this object is now based on a plaintext key, not a stored CNG key.
+        }
+
+        public void GenerateKey()
+        {
+            byte[] key = new byte[_outer.BaseKeySize.BitSizeToByteSize()];
+            do
+            {
+                RandomNumberGenerator.Create().GetBytes(key);
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: RandomNumberGenerator is IDisposable. We should Dispose of the instance when done with it, even if the current implementations have Dispose as a nop.

File: src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            }
+            while (_outer.IsWeakKey(key));
+            SetKey(key);
+        }
+
+        public void GenerateIV()
+        {
+            byte[] iv = new byte[_outer.BlockSize.BitSizeToByteSize()];
+            RandomNumberGenerator.Create().GetBytes(iv);
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: Ditto.

File: src/System.Security.Cryptography.Cng/src/Internal/Cryptography/SymmetricImportExportExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                    bw.Write((char)0);
+
+                    // Write out a BCRYPT_KEY_DATA_BLOB_HEADER
+                    bw.Write((int)Interop.BCrypt.BCRYPT_KEY_DATA_BLOB_MAGIC);          // BCRYPT_KEY_DATA_BLOB_HEADER.dwMagic
+                    bw.Write((int)Interop.BCrypt.BCRYPT_KEY_DATA_BLOB_VERSION1);       // BCRYPT_KEY_DATA_BLOB_HEADER.dwVersion
+                    bw.Write((int)(key.Length));                                       // BCRYPT_KEY_DATA_BLOB_HEADER.cbKeyData
+
+                    bw.Write((byte[])key);                                             // Write out the key data.
+                }
+                ms.Flush();
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/SymmetricImportExportExtensions.cs]: MemoryStream.Flush is a nop

File: src/System.Security.Cryptography.Cng/src/Internal/Cryptography/SymmetricImportExportExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+using System.Globalization;
+using System.Security.Cryptography;
+
+namespace Internal.Cryptography
+{
+    internal static class SymmetricImportExportExtensions
+    {
+        public static CngKey ToCngKey(this byte[] key, string algorithm)
+        {
+            using (MemoryStream ms = new MemoryStream())
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/SymmetricImportExportExtensions.cs]: Consider passing in the expected capacity here so that the MemoryStream doesn't need to continually grow its array.

File: src/Common/src/Interop/Windows/Crypt32/Interop.NCryptAlgorithms.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class Crypt32
```
Comment (bartonjs) [inline src/Common/src/Interop/Windows/Crypt32/Interop.NCryptAlgorithms.cs]: Wouldn't NCrypt algorithms be in NCrypt, not Crypt32?

File: src/Common/tests/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        [Fact]
+        public static void ZeroLengthTransformNoPad()
+        {
+            // Use no TransformBlock calls() - do the entire transform using only TransformFinalBlock().
+            byte[] key = "1ed2f625c187b993256a8b3ccf9dcbfa5b44b4795c731012f70e4e64732efd5d".HexToByteArray();
+            byte[] iv = "47d1e060ba3c8643f9f8b65feeda4b30".HexToByteArray();
+
+            byte[] plainText = "".HexToByteArray();
+            // CBC, Padding.None
+            byte[] expectedCipher = "".HexToByteArray();
```
Comment (bartonjs) [inline src/Common/tests/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: Why not just Array.Empty\<byte\>?

File: src/Common/src/Interop/Windows/Crypt32/Interop.NCryptAlgorithms.cs (hunk 6 above)
Comment (AtsushiKan) [inline src/Common/src/Interop/Windows/Crypt32/Interop.NCryptAlgorithms.cs]: Perhaps but that follows existing naming pattern in Common\Src\Interop.

File: src/Common/tests/Cryptography/AlgorithmImplementations/TripleDES/TripleDESCipherTests.cs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+using Xunit;
+
+namespace System.Security.Cryptography.Encryption.TripleDes.Tests
+{
+    public static class TripleDESCipherTests
+    {
+        [Fact]
+        public static void TripleDESDefaults()
+        {
+            TripleDES des = TripleDESFactory.Create();
```
Comment (bartonjs) [inline src/Common/tests/Cryptography/AlgorithmImplementations/TripleDES/TripleDESCipherTests.cs]: TripleDESRoundTrip... uses a using, why doesn't this?

File: src/Common/src/Interop/Windows/Crypt32/Interop.NCryptAlgorithms.cs (hunk 6 above)
Comment (AtsushiKan) [inline src/Common/src/Interop/Windows/Crypt32/Interop.NCryptAlgorithms.cs]: On second thought, I think I’m confused. Wait a sec…

File: src/Common/tests/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs (hunk 7 above)
Comment (AtsushiKan) [inline src/Common/tests/Cryptography/AlgorithmImplementations/AES/AesCornerTests.cs]: Symmetry with the other crytotests that use the <string>.ToHexByteArray() pattern.

File: src/Common/tests/Cryptography/CryptoUtils.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+using System.Globalization;
+using System.Security.Cryptography;
+
+namespace Test.Cryptography
+{
+    internal static class CryptoUtils
+    {
+        internal static byte[] Encrypt(this SymmetricAlgorithm alg, byte[] plainText, int blockSizeMultipler = 1)
+        {
+            return alg.CreateEncryptor().Transform(plainText, blockSizeMultipler);
```
Comment (bartonjs) [inline src/Common/tests/Cryptography/CryptoUtils.cs]: the output of CreateEncryptor is disposable, and should be in a using.

File: src/Common/tests/Cryptography/CryptoUtils.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        internal static byte[] Decrypt(this SymmetricAlgorithm alg, byte[] cipher, int blockSizeMultipler = 1)
+        {
+            return alg.CreateDecryptor().Transform(cipher, blockSizeMultipler);
```
Comment (bartonjs) [inline src/Common/tests/Cryptography/CryptoUtils.cs]: the output of CreateDecryptor is disposable, and should be in a using.

File: src/Common/tests/Cryptography/CryptoUtils.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                if (count >= blockSize)
+                {
+                    byte[] buffer = new byte[blockSize];
+                    int numBytesWritten = transform.TransformBlock(input, i, count, buffer, 0);
+                    Array.Resize(ref buffer, numBytesWritten);
+                    output.AddRange(buffer);
+                }
+                else
+                {
+                    byte[] finalBlock = transform.TransformFinalBlock(input, i, count);
```
Comment (bartonjs) [inline src/Common/tests/Cryptography/CryptoUtils.cs]: I guess I don't understand what blockSizeMultiplier is supposed to be doing.

Transform("hi".ToBytes(), 10) is the same as Transform("hi".ToBytes(), 1), in that both just call TransformFinalBlock.

And since blocksize isn't decremented during looping, Transform(new byte[4096], 1) is just going to call TransformBlock over and over, but never TransformFinalBlock.

Comment (AtsushiKan) [inline src/Common/tests/Cryptography/CryptoUtils.cs]: “I” is the loop induction variable, not “blockSize”.

The multiplier controls the size of the chunks we feed to Transform(). Purpose is to let us test multiple block transforms, and edge cases (like never calling Transform but only TransformFinalBlock.)

File: src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+                {
+                    return cngKey.GetSymmetricKeyDataIfExportable(_algorithm);
+                }
+            }
+        }
+
+        public void SetKey(byte[] key)
+        {
+            _outer.BaseKey = key;
+            _keyName = null; // Setting _keyName to null signifies that this object is now based on a plaintext key, not a stored CNG key.
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: Opening an AesCng with a named key, then assigning the Key property switches to a software ephemeral key.  Is that what we'd expect, or should it just throw?

Comment (AtsushiKan) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: Fair question. I’m no fan of the way these Crypto classes let you reassign everything under the sun instead of specifying things once and for all in a constructor.

On the other hand, that is the pattern, so I didn’t want to change it for this one case.

Comment (bartonjs) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: Well, the existing things let you change a software key to another software key.  In this case, it feels like it should either
a) Update the key in the named key container
b) throw

But
c) switch from a saved (and maybe hardware) key to an unsaved software key

doesn't seem very good.  I'm a fan of b; personally.

Comment (AtsushiKan) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: I would definitely not vote for (a). Exchanging one software key to another doesn’t propagate the change backward to any other component that happened to be storing the old key. Nor would I expect it to.

(b) Throwing a “Make up your mind” exception if you attempt to set Key or KeySize after constructing the SA with a named key is fine by me
      philosophically, but it isn’t consistent with the overall flavor of SymmetricAlgorithm-based classes so I’d like to get an opinion from the api guys here.

Comment (AtsushiKan) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: cc @nguerrera 

Do you have an opinion on this?

Comment (bartonjs) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: @nguerrera Here's the summary:

```
using (Aes maybeHardwareAes = new AesCng(keyAddress, keyProvider))
{
    // What should this line do?
    maybeHardwareAes.Key = new byte[128 / 8];
}
```

Comment (nguerrera) [inline src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs]: Looked at this a bit offline with @bartonjs. We already have the same situation in the RSA/symmetric case, where you can create an RSACng from a hardware key and then change the key to a new software one using KeySize setter or Import. We should be consistent so I think the way @AtsushiKan is fine unless you guys want to make an all-up proposal for changing the behavior elsewhere too.

**Comment by bartonjs** [comment]
LGTM

**Comment by stephentoub** [comment]
LGTM
